### PR TITLE
Add system.storage.Region support

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

**Review by floitsch — APPROVED** [review]
LGTM.

File: lib/system/storage.toit [hunk 1]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+    erase_granularity := 1 << reply[3]
+    erase_granularity_mask_ = erase_granularity - 1
+    size = reply[2]
+    super client handle
+    try:
+      resource_ = flash_region_open_
+          resource_freeing_module_
+          client.id
+          handle
+          reply[1]
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:165]: Done.

File: lib/system/storage.toit [hunk 2]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+  /**
+  Opens a storage region using the given $scheme and $path.
+
+  If an existing region matches the $scheme and $path, it
+    is opened. An exception is thrown if a $capacity is
+    provided and the existing region is smaller than that.
+
+  If no region that match $scheme and $path exists, a new
+    one is created. An exception is thrown if no $capacity
+    is provided.
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:240]: I don't think we're very stringent with the distinction between exception and errors yet. But done.

File: lib/system/storage.toit [hunk 3]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+  /**
+  Opens a storage region using the $SCHEME_FLASH scheme and the
+    given $path.
+
+  See $(open --scheme --path --capacity) for an explanation of
+    the other parameters.
+  */
+  static open --flash/bool path/string -> Region
+      --capacity/int?=null:
+    if flash != true: throw "Bad Argument"
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:228]: Done.

File: src/flash_allocation.cc [hunk 4]
```diff
@@ -20,56 +20,68 @@
...
-}
-
-void FlashAllocation::Header::set_uuid(const uint8* uuid) {
-  memmove(uuid_, uuid, UUID_SIZE);
-}
-
-void FlashAllocation::validate() {
-  // Do nothing.
+static void initialize(void* dst, const void* src, size_t size) {
+  if (src) {
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/flash_allocation.cc:35]: We use this kind of null-check in a lot of code (mostly done by Erik), but done.

File: src/flash_allocation.h [hunk 5]
```diff
@@ -22,107 +22,78 @@
...
-      memcpy(id_, id, id_size());
-      pages_in_flash_ = static_cast<uint16>(Utils::round_up(size, FLASH_PAGE_SIZE) >> 12);
-      memcpy(_metadata, metadata, METADATA_SIZE);
-      set_uuid(uuid);
-    }
-
-    static int id_size() { return sizeof(id_); }
     const uint8* id() const { return id_; }
-    int size() const { return pages_in_flash_ << 12; }
+    int size() const { return size_in_pages_ << 12; }
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/flash_allocation.h:45]: Done.

File: src/flash_registry_shared.cc [hunk 6]
```diff
@@ -27,59 +28,76 @@ int FlashRegistry::find_next(int offset, ReservationList::Iterator* it) {
...
-    return const_cast<char*>(allocations_memory()) + offset;
+static bool is_erased_aligned(const uint8* memory, int from, int to) {
+  ASSERT(Utils::is_aligned(memory + from, sizeof(uword)));
+  ASSERT(Utils::is_aligned(memory + to, sizeof(uword)));
+  const uword* cursor = reinterpret_cast<const uword*>(memory + from);
+  const uword* limit = reinterpret_cast<const uword*>(memory + to);
+  do {
+    if (*cursor != static_cast<uword>(-1)) return false;
+    ++cursor;
+  } while (cursor < limit);
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/flash_registry_shared.cc:76]: Done (I copied this from the ESP32 code).

File: src/primitive_flash_registry.cc [hunk 7]
```diff
@@ -158,4 +153,160 @@ PRIMITIVE(erase_flash_registry) {
   return BOOL(FlashRegistry::erase_flash_registry());
 }
 
+PRIMITIVE(allocate) {
+  PRIVILEGED;
+  ARGS(int, offset, int, size, int, type, Blob, id, Blob, metadata, Blob, content);
+  for (auto it = reservations.begin(); it != reservations.end(); ++it) {
+    Reservation* reservation = *it;
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/primitive_flash_registry.cc:160]: Done.

File: src/primitive_flash_registry.cc [hunk 8]
```diff
@@ -158,4 +153,160 @@ PRIMITIVE(erase_flash_registry) {
...
+}
+
+PRIMITIVE(grant_access) {
+  PRIVILEGED;
+  ARGS(int, client, int, handle, int, offset, int, size);
+  RegionGrant* grant = _new RegionGrant(client, handle, offset, size);
+  if (!grant) MALLOC_FAILED;
+  Locker locker(OS::global_mutex());
+  for (auto it : grants) {
+    if (it->offset() == offset && it->size() == size) ALREADY_IN_USE;
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/primitive_flash_registry.cc:198]: It is just to get an error thrown if you open the same flash region twice. It is covered by tests.

File: src/primitive_flash_registry.cc [hunk 9]
```diff
@@ -158,4 +153,160 @@ PRIMITIVE(erase_flash_registry) {
...
+PRIMITIVE(region_close) {
+  ARGS(FlashRegion, resource);
+  resource->resource_group()->unregister_resource(resource);
+  resource_proxy->clear_external_address();
+  return process->program()->null_object();
+}
+
+static bool is_within_bounds(FlashRegion* resource, int from, int size) {
+  int to = from + size;
+  return from >= 0 && to > from && to <= resource->size();
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/primitive_flash_registry.cc:272]: Done.

File: lib/system/storage.toit [hunk 10]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+  is_erased --from/int=0 --to/int=size -> bool:
+    if not resource_: throw "ALREADY_CLOSED"
+    return flash_region_is_erased_ resource_ from (to - from)
+
+  /**
+  Erases the bytes in the range starting at $from and ending
+    at $to (exclusive).
+
+  Throws an exception if the $from and $to parameters must
+    both be aligned to $erase_granularity.
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:309]: Done.

File: lib/system/storage.toit [hunk 11]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+
+  If no exception is thrown, the bytes in the specified
+    range are all set to $erase_value.
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:312]: Done.

File: lib/system/storage.toit [hunk 12]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+  If no region that match $scheme and $path exists, a new
+    one is created. An exception is thrown if no $capacity
+    is provided.
+  */
+  static open --scheme/string --path/string -> Region
+      --capacity/int?=null:
+    client := _client_
+    if not client: throw "UNSUPPORTED"
+    if capacity and capacity < 1: throw "Bad Argument"
+    path.index_of ":" --if_absent=:
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:247]: Done.

File: lib/system/storage.toit [hunk 13]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+  See $(open --scheme --path --capacity) for an explanation of
+    the other parameters.
+  */
+  static open uri/string -> Region
+      --capacity/int?=null:
+    split := uri.index_of ":" --if_absent=: throw "No scheme provided"
+    return open --scheme=uri[..split] --path=uri[split + 1 ..] --capacity=capacity
+
+  /**
+  Opens a storage region using the $SCHEME_FLASH scheme and the
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:220]: Done.

File: lib/system/storage.toit [hunk 14]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+
+  /**
+  Whether a call to $write can change individual bits
+    from 1 to 0.
+  */
+  write_can_clear_bits -> bool:
+    return (mode_ & MODE_WRITE_CAN_CLEAR_BITS_) != 0
+
+  /**
+  Opens a storage region with using the schema and path parsed
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:204]: Done.

File: lib/system/storage.toit [hunk 15]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+    from the given $uri.
+
+  The format of the $uri is <scheme>:<path> and it is common to
+    use qualified paths that include the domain name of the
+    region owner, e.g. "flash:toitlang.org/jag".
+
+  See $(open --scheme --path --capacity) for an explanation of
+    the other parameters.
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:212]: Done.

File: lib/system/storage.toit [hunk 16]
```diff
@@ -95,7 +102,243 @@ class Bucket extends ServiceResourceProxy:
...
+      --capacity/int?=null:
+    split := uri.index_of ":" --if_absent=: throw "No scheme provided"
+    return open --scheme=uri[..split] --path=uri[split + 1 ..] --capacity=capacity
+
+  /**
+  Opens a storage region using the $SCHEME_FLASH scheme and the
+    given $path.
+
+  See $(open --scheme --path --capacity) for an explanation of
+    the other parameters.
```
[the remark this answers is not in the document]
Comment (kasperl) [reply lib/system/storage.toit:224]: Done.

**Pull request merged**